### ISSUE TITLE
Allow ldap auth via rest api

### ISSUE THREAD

**Issue by issuemover631** [airsonic/airsonic#260]
<a href="https://github.com/muff1nman"><img src="https://avatars2.githubusercontent.com/u/1531528?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [muff1nman](https://github.com/muff1nman)**
_Saturday Feb 25, 2017 at 23:47 GMT_
_Originally opened as https://github.com/Libresonic/libresonic/issues/258_

----

When I was fixing up the ldap auth a week or so back I noticed that the REST api stuff uses a different authentication mechanism which makes it so that ldap authentication does not work via the REST api and by extension android apps. We should really change the auth mechanism on the REST endpoints to use the same auth as the rest of the app. This might be work that should be done along with #69.

**Comment by muff1nman** [comment]
Verified this is an issue today

**Comment by thallian** [comment]
For some reason the DSub Variant from F-Droid works for me with LDAP auth. None of the other apps though.

**Comment by kykc** [comment]
Strangely, I can confirm @thallian statement: D-Sub from F-Droid works fine with airsonic 10.1.2 and LDAP, but D-Sub from play market doesn't

**Comment by kykc** [comment]
Yep, that's the reason. I've compiled 5.4.2 floss version from sources and it's not working. If I will have enough spare time I'll try to pinpoint exact version of D-Sub when this stopped working and peek and the diff involved, maybe I'll get some insights regarding the issue.

**Comment by hjone72** [comment]
I've narrowed the cause down to a change between v5.2.2 and v5.3 of DSub. Hope that helps...
https://github.com/daneren2005/Subsonic/compare/5.2.2...5.3

**Comment by muff1nman** [comment]
cc @daneren2005

**Comment by PoGo606** [comment]
Same problem here, can't connect with Dsub from Play Store and LDAP auth configured...

**Comment by hjone72** [comment]
@PoGo606 The problem can be found here: https://github.com/popeen/Popeens-DSub/issues/68#issuecomment-494649972

I'm not sure how to fix it however I used nginx subfilter to change the response and got it working.

```
        sub_filter_types text/xml application/json;
        sub_filter_once off;
        sub_filter 'subsonic' 'madsonic';
```

**Comment by PoGo606** [comment]
Thank you @hjone72 for your answer.

Actually, I think I have a different issue here.
Local and LDAP accounts are KO with Dsub but OK when using Subsonic official app.
Looking at HTTP requests/response it seems the issue is from the API auth mecanism used by Dsub.

Here is the request made by **DSUB App**:
`GET /airsonic/rest/ping.view?u=pogo&s=vqndtpmq3m6tecpok1gdfnb0br&t=66314741d0d8ae5e85d3362897957e3f&v=1.2.0&c=DSub HTTP/1.1`

Response:
`<?xml version="1.0" encoding="UTF-8"?>
<subsonic-response xmlns="http://subsonic.org/restapi" status="failed" version="1.15.0">
   <error code="40" message="Wrong username or password."/>
</subsonic-response>`

Same with **official Subsonic app**:
`POST /airsonic/rest/ping.view?u=pogo&p=enc:6e79756d617469636174696f6e&v=1.2.0&c=android HTTP/1.1`

Response:
`<?xml version="1.0" encoding="UTF-8"?>
<subsonic-response xmlns="http://subsonic.org/restapi" status="ok" version="1.15.0"/>`

Both App are working well with official Subsonic server.
So it seems Airsonic doesn't handle well the first auth mecanism (GET + Encrypted + Salted ?)

**Comment by hjone72** [comment]
There is GET request that the app does to find out what capabilities the server has. If the server returns 'subsonic' the app thinks the server supports (GET + Encrypted + Salted). When the server responds with 'madsonic' it uses the old method.

It was a while ago that I found these bits and pieces so I'm just working from memory here. I believe it could be fixed either at the app end or the server end. I didn't have time to figure it out so just put a dirty hack in place. :)

**Comment by dvdkon** [comment]
The challenge-response mechanism of sending `md5(password + nonce)` requires knowing the plain-text password. However, Airsonic's LDAP auth works by getting the plain-text password from the user and trying to authenticate ("bind") to the LDAP server with it, so this challenge-response scheme can't work with it, or any reasonable external auth mechanism.

This isn't so tragic. Sending plain-text passwords over HTTPS is arguably much more secure than storing them in a database, and everyone should be using HTTPS by now, so I'd recommend deprecating this authentication mechanism. Hopefully it will make at least some clients will switch to the other, ironically more modern, method.

**Comment by loki666** [comment]
what are you suggesting to deprecate ? LDAP auth or md5(password + nonce) ?

**Comment by dvdkon** [comment]
Sorry I wasn't clear, the `md5(password + nonce)` one.

**Comment by jmccoy555** [comment]
> @PoGo606 The problem can be found here: [popeen/Popeens-DSub#68 (comment)](https://github.com/popeen/Popeens-DSub/issues/68#issuecomment-494649972)
> 
> I'm not sure how to fix it however I used nginx subfilter to change the response and got it working.
> 
> ```
>         sub_filter_types text/xml application/json;
>         sub_filter_once off;
>         sub_filter 'subsonic' 'madsonic';
> ```

Does the trick, just add `--with-http_sub_module \` if you build Nginx

**Comment by mmguero** [comment]
Having the same problem. :(

**Comment by frdbonif** [comment]
> > @PoGo606 The problem can be found here: [popeen/Popeens-DSub#68 (comment)](https://github.com/popeen/Popeens-DSub/issues/68#issuecomment-494649972)
> > I'm not sure how to fix it however I used nginx subfilter to change the response and got it working.
> > ```
> >         sub_filter_types text/xml application/json;
> >         sub_filter_once off;
> >         sub_filter 'subsonic' 'madsonic';
> > ```
> 
> Does the trick, just add `--with-http_sub_module \` if you build Nginx

An old issue, I know.  I've just got everything up and running with LDAPS authentication but can't get DSub authenticating.  Can you or anyone else clarify where I put this into the Nginx configuration?

I guess that in reality this is an issue that needs looking at within the DSub code itself?
